include/oneapi/dpl/internal/dynamic_selection_impl/concurrent_queue.h

```diff
@@ -0,0 +1,76 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) 2023 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+
+#ifndef _ONEDPL_CONCURRENT_QUEUE_H
+#define _ONEDPL_CONCURRENT_QUEUE_H
+
+#include <queue>
+#include <mutex>
+#include <list>
+#include <condition_variable>
+
+namespace oneapi {
+namespace dpl{
+namespace experimental{
+namespace util{
+    template <typename T>
+    class concurrent_queue
+    {
+    public:
+
+        void pop(T& item)
+        {
+            std::unique_lock<std::mutex> mlock(mutex_);
+            while (queue_.empty())
+            {
+                cond_.wait(mlock);
+            }
+            item = queue_.front();
+            queue_.pop();
+        }
+
+        void push(T item)
+        {
+            std::unique_lock<std::mutex> mlock(mutex_);
+            queue_.push(item);
+            mlock.unlock();
+            cond_.notify_one();
+        }
+
+        void pop_all(std::list<T>& item_list)
+        {
+            std::unique_lock<std::mutex> mlock(mutex_);
+            while(!queue_.empty()){
+                auto item = queue_.front();
+                queue_.pop();
+                item_list.push_back(item);
+
+            }
+        }
+
+        bool is_empty(){
+            return queue_.empty();
+        }
+
+        concurrent_queue() = default;
+        concurrent_queue(const concurrent_queue& q) : queue_(q.queue_) {};
+        concurrent_queue& operator=(const concurrent_queue& q) {q=q.queue_; return *this;}
```

Comment:
I think the source queue needs to be locked when you make a copy of it. If it's OK for queue copying to be thread-unsafe for the source, at least add a comment here.

Reply:
@AnuyaWelling2801 can you check if we need to support copying of concurrent_queue. My suspicion is that it is only used in schedulers and they themselves are not copyable and so their internal queues do not need to be copyable.

Reply:
Yes, good point. We can delete the copy constructor.